expression/builtin_cast.go

```diff
@@ -461,9 +461,10 @@ func (b *builtinCastIntAsRealSig) evalReal(row chunk.Row) (res float64, isNull b
 	if isNull || err != nil {
 		return res, isNull, err
 	}
-	if !mysql.HasUnsignedFlag(b.tp.Flag) && !mysql.HasUnsignedFlag(b.args[0].GetType().Flag) {
+	unsigned := mysql.HasUnsignedFlag(b.args[0].GetType().Flag)
+	if !mysql.HasUnsignedFlag(b.tp.Flag) && !unsigned {
 		res = float64(val)
-	} else if b.inUnion && val < 0 {
+	} else if b.inUnion && !unsigned && val < 0 {
```

Comment:
I think `} else if b.inUnion && mysql.HasUnsignedFlag(b.tp.Flag) && val < 0 {` is more comprehensable.